odoo/addons/test_lint/tests/_odoo_checker_gettext.py

```diff
@@ -49,10 +50,46 @@ class OdooBaseChecker(BaseChecker):
             'gettext-repr',
             'Don\'t use %r to automatically insert quotes in translation strings. Quotes can be different depending on the language: they must be part of the translated string.',
         ),
+        'E8507': (
+            'Static string passed to %s without gettext call.',
+            'missing-gettext',
+            'Ensure that all static strings passed to certain constructs are wrapped in a gettext call.',
+        ),
     }
 
-    @only_required_for_messages('gettext-variable', 'gettext-placeholders', 'gettext-repr')
+    required_inside_functions = ['UserError', 'ValidationError', 'AccessError', 'AccessDenied', 'MissingError']
+
+    @only_required_for_messages('missing-gettext', 'gettext-variable', 'gettext-placeholders', 'gettext-repr')
     def visit_call(self, node):
+        file_path = self.linter.current_file
+        if (
+            "test" in file_path
+            and (module_path := find_module_root(file_path))
+            and (
+                # Check if the module's base name starts with "test_"
+                # This identifies modules dedicated to testing (e.g., test_mail)
+                os.path.basename(os.path.normpath(module_path)).startswith("test_")
+                # OR check if the file is in a "tests" subdirectory within the module
+                # This identifies test files located within the standard "tests" directory/package
+                or file_path[len(module_path) + 1 :].startswith("tests" + os.sep)
+            )
+        ):
+            return
+
+        if (
+            isinstance(node.func, astroid.Name)
+            and node.func.name in self.required_inside_functions
+            and len(node.args) > 0
+        ):
+            first_arg = node.args[0]
+            if (
+                isinstance(first_arg, astroid.Const)
+                and isinstance(first_arg.value, str)
+                and not (isinstance(first_arg.parent, astroid.Call) and first_arg.parent.func.name in ("_", "_lt"))
```

Comment:
I'm a bit confused about what `first_arg.parent` is. I'm not sure this behaves as you'd expect. I think this line is useless: the parent of the child of `UserError` is `UserError`, it will never be `_` or `_lt`.

I think the only two things we want to accept inside UserError are:
1. `astroid.Name` (in case they pass a variable. Note: what about `astroid.Attribute`?)
2. `astroid.Call` (only if it's a call to gettext)
Everything else should produce an error

Reply:
I think we should also include `astroid.Attribute` because `_` could be used as `odoo.translate._` for example?

Reply:
I think `astroid.Call` covers both `astroid.Attribute` and `astroid.Name` (both methods and unqualified function names). My thought about whitelisting `astroid.Attribute` was because someone could provide `self.some_translated_content` as an argument to `UserError`, and we wouldn't want to prevent them to do so